api/filters/imagetag/updater.go

```diff
@@ -30,14 +30,25 @@ func (u imageTagUpdater) SetImageValue(rn *yaml.RNode) error {
 		return nil
 	}
 
-	name, tag := image.Split(value)
+	name, tag, digest := image.Split(value)
 	if u.ImageTag.NewName != "" {
 		name = u.ImageTag.NewName
 	}
-	if u.ImageTag.NewTag != "" {
-		tag = ":" + u.ImageTag.NewTag
+	// not overriding tag/digest component, keep original
+	if u.ImageTag.NewTag == "" && u.ImageTag.Digest == "" {
+		if tag != "" {
+			tag = ":" + tag
+		}
+		if digest != "" {
+			tag += "@" + digest
+		}
 	}
-	if u.ImageTag.Digest != "" {
+	// overriding tag or digest will replace both original tag and digest values
+	if u.ImageTag.NewTag != "" && u.ImageTag.Digest != "" {
+		tag = ":" + u.ImageTag.NewTag + "@" + u.ImageTag.Digest
+	} else if u.ImageTag.NewTag != "" {
+		tag = ":" + u.ImageTag.NewTag
+	} else if u.ImageTag.Digest != "" {
```

Comment:
suggest changing line 37-55 to something such as

```
if u.ImageTag.NewTag != "" {
  tag = u.ImageTag.NewTag
}
if u.ImageTag.Digest != "" {
  digest = u.ImageTag.Digest
}
if tag != "" {
  name = name + ":" + tag
}
if digest != "" {
  name = name + "@" + tag
}
return u.trackableSetter.SetScalar(name)(rn)
```

It reads a bit simpler.

Reply:
If the original image has both tag and digest and the image filter only specifies `newTag` or `digest` I think the expectation would be to override both. I'm not sure it makes much sense to keep the original digest or tag components when the filter overrides either of them.

For examples given the  original image`nginx:1.2.3@sha256:xyz`:
- only `newTag: 2.0.0` set -> `nginx:2.0.0`
- only `digest: sha256:abc` set-> `nginx@sha256:abc`
- `newTag: 2.0.0` and `digest: sha256:abc` set -> `nginx:2.0.0@sha256:abc`

I think the changes you are suggesting would result in the following:
- only `newTag: 2.0.0` set -> `nginx:2.0.0@sha256:xyz`
- only `digest: sha256:abc` set-> `nginx:1.2.3@sha256:abc`
- `newTag: 2.0.0` and `digest: sha256:abc` set -> `nginx:2.0.0@sha256:abc`

In addition there would be no way to unset the original tag or digest values.

I refactored it with some of your suggestions to simplify the string builder, but kept the original behavior.

Reply:
Ah thank you, that makes sense. I hadn’t thought through the implication of keeping the original in the case that only one is overridden.